Feature/tlvf types

tomereli opened the pull request:
Add tlvftypes.h, which will be used as a container header for all the
tlvf defined types, and in addition include all the operator overloads,
thus allowing a single header include by users of the tlvf library for
using a certain type.

Move all the operator overloads of sMacAddr which are now defined in
BCL to tlvftypes.h. This will now allow an external application to link
to TLVF alone for using sMacAddr effectively, the first "customer" being
the transport.

Commit:
tlvf: run formatting on tlvf.py

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: remove tlvfutils.h

tlvfutils.h is empty, remove it and all of its references throuout the
code.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: add tlvftypes

Add tlvftypes.h and tlvftypes.cpp, which will be used as a container  for all the
tlvf defined types, and in addition include all the operator overloads,
thus allowing a single header include by users of the tlvf library for
using a certain type.

Move all the operator overloads of sMacAddr which are now defined in
BCL to tlvftypes.h. This will now allow an external application to link
to TLVF alone for using sMacAddr effectively, the first "customer" being
the transport.

Move all the mac_to_string and mac_from_string utility functions from
bcl::network_utils.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
transport: Use sMacAddr

Now that sMacAddr is part of tlvf, link the transport to tlvf and
replace any use of mac address with sMacAddr.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

morantr commented on common/beerocks/bcl/source/network/network_utils.cpp:
```diff
@@ -84,59 +84,27 @@ const std::string network_utils::MULTICAST_1905_MAC_ADDR("01:80:c2:00:00:13");
 std::string network_utils::mac_to_string(const sMacAddr &mac)
 {
-    return mac_to_string((const uint8_t *)mac.oct);
+    return tlvf::mac_to_string((const uint8_t *)mac.oct);
```
I think it is better to remove these functions from network utils, and add one more commit that replaces all the calling for these functions to have the `tlvf::` prefix instead of `network_utils::`.

tomereli replied:
You just love these huge changes hu? :)
OK Done, but I really hope it will be merged soon otherwise it will be a nightmare on rebases.

morantr commented on framework/transport/ieee1905_transport/ieee1905_transport_packet_processing.cpp:
```diff
@@ -541,18 +541,18 @@ std::ostream &Ieee1905Transport::Packet::print(std::ostream &os) const
+       << std::hex << std::setfill('0') << std::setw(2) << (unsigned)src.oct[2] << ":" << std::hex
+       << std::setfill('0') << std::setw(2) << (unsigned)src.oct[3] << ":" << std::hex
+       << std::setfill('0') << std::setw(2) << (unsigned)src.oct[4] << ":" << std::hex
+       << std::setfill('0') << std::setw(2) << (unsigned)src.oct[5] << std::endl;
```
You don't need to do that. We already have operator overload for this. If you do ` << dst` it will be converted to a string.

tomereli replied:
right, fixed.

morantr reviewed (approved):
It makes me wonder. A lot of code on BCL may be fit more to the framework and can be used there.
string utils for example, too bad that we are duplicating code for that.

tomereli commented:
> It makes me wonder. A lot of code on BCL may be fit more to the framework and can be used there.
> string utils for example, too bad that we are duplicating code for that.

You are right, but I think it is OK to move things in steps when they are needed. I needed sMacAddr so I used this opportunity to move it now.

mariomaz commented on framework/tlvf/src/src/tlvftypes.cpp:
```diff
@@ -0,0 +1,81 @@
+namespace tlvf {
+std::string int_to_hex_string(const unsigned int integer, const uint8_t number_of_digits)
+{
+    // 'number_of_digits' represent how much digits the number should have, so the function will
```
These comments should be moved to header file, inside a properly doxygen-formatted documentation

tomereli replied:
Done.

mariomaz commented on framework/tlvf/src/src/tlvftypes.cpp:
```diff
@@ -0,0 +1,81 @@
+{
+    std::string mac_addr_string;
+
+    mac_addr_string = int_to_hex_string((uint32_t)mac_address[0], 2) + ":" +
```
Define a constexpr "separator" character instead of repeating ":" all along this file

tomereli replied:
I don't really see the benefit for doing this cause it is used only in two functions, once as char * and once as a char, but OK.

mariomaz commented on framework/tlvf/src/src/tlvftypes.cpp:
```diff
@@ -0,0 +1,81 @@
+                      int_to_hex_string((uint32_t)mac_address[2], 2) + ":" +
+                      int_to_hex_string((uint32_t)mac_address[3], 2) + ":" +
+                      int_to_hex_string((uint32_t)mac_address[4], 2) + ":" +
+                      int_to_hex_string((uint32_t)mac_address[5], 2);
```
Write a loop instead of repeating 6 times the same line

tomereli replied:
Its not the same line, the last one doesn't include the `":"` separator.

mariomaz replied:
I meant this:
```c
for (size_t i = 0; i < ETHA_LEN; i++) {
    if (i > 0) {
        mac_addr_string += mac_bytes_separator;
    }
    mac_addr_string += int_to_hex_string(mac_address[i], 2);
}

mariomaz commented on framework/tlvf/src/src/tlvftypes.cpp:
```diff
@@ -0,0 +1,81 @@
+    uint8_t mac_address[ETH_ALEN];
+    int8_t i;
+    uint8_t *p = mac_address;
+    for (i = 5; i >= 0; i--) {
```
Use ETH_ALEN instead of magic number

tomereli replied:
ETH_ALEN is 6.

mariomaz replied:
I meant this:
```c
for (i = ETH_ALEN - 1; i >= 0; i--) {

mariomaz commented on framework/tlvf/src/src/tlvftypes.cpp:
```diff
@@ -0,0 +1,81 @@
+void mac_from_string(uint8_t *buf, const std::string &mac)
+{
+    if (mac.empty()) {
+        memset(buf, 0, ETH_ALEN);
```
"buf" might be null

Commit:
bcl: remove mac_to_string and mac_from_string wrappers

Now that these APIs are part of TLVF, remove the wrappers
in the BCL, and modify the rest of the code to use the tlvf functions
instead of the bcl ones.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

mariomaz commented on framework/transport/ieee1905_transport/include/mapf/transport/ieee1905_transport.h:
```diff
@@ -143,8 +144,8 @@ class Ieee1905Transport {
         unsigned int src_if_index = 0;
-        uint8_t dst[ETH_ALEN]     = {0}; // destination mac address
-        uint8_t src[ETH_ALEN]     = {0}; // source mac address
+        sMacAddr dst              = {.oct = {0}}; // destination mac address
```
Maybe you should also move ZERO_MAC from network_utils to TLVF

tomereli replied:
True, but there if I do this I can also move a lot more. I prefer to keep this for a separate PR.

mariomaz commented on framework/transport/ieee1905_transport/include/mapf/transport/ieee1905_transport.h:
```diff
@@ -220,7 +221,7 @@ class Ieee1905Transport {
         {
             // compare based on src address first then messageId
-            int srccmp = memcmp(lhs.src, rhs.src, ETH_ALEN);
+            int srccmp = memcmp(lhs.src.oct, rhs.src.oct, ETH_ALEN);
```
Now you can use operator ==

tomereli replied:
I prefer to do this as well in a separate PR since it is changing the logic.

mariomaz reviewed (approved):
A few comments but approving, you decide
